be/src/runtime/batch_write/txn_state_cache.cpp

```diff
@@ -0,0 +1,467 @@
+    if (_stopped) {
+        return;
+    }
+    _stopped = true;
+    _cv.notify_all();
+}
+
+std::string TxnStateHandler::debug_string() {
+    std::unique_lock<bthread::Mutex> lock(_mutex);
+    return fmt::format(
+            "txn_id: {}, txn_status: {}, reason: {}, num_subscriber: {}, num_waiting_finished_state: {}, stopped: {}",
+            _txn_id.load(), to_string(_txn_state.txn_status), _txn_state.reason, _num_subscriber,
+            _num_waiting_finished_state, _stopped);
+}
+
+void TxnStateHandler::_transit_txn_state(TTransactionStatus::type new_status, const std::string& reason, bool from_fe) {
+    TTransactionStatus::type old_status = _txn_state.txn_status;
+    TRACE_BATCH_WRITE << "receive new txn state, txn_id: " << _txn_id
+                      << ", current status: " << to_string(_txn_state.txn_status)
+                      << ", current reason: " << _txn_state.reason << ", new status: " << new_status
+                      << ", new reason: " << reason << ", from_fe: " << from_fe;
+    // special case for COMMITTED status. If it's notified by FE, it means the load finished with
+    // publish timeout, _is_finished_txn_state() should return true, and notify subscribers.
+    if (new_status == TTransactionStatus::COMMITTED && from_fe) {
+        _committed_status_from_fe = from_fe;
+    }
+    if (old_status == TTransactionStatus::VISIBLE || old_status == TTransactionStatus::ABORTED ||
+        old_status == TTransactionStatus::UNKNOWN) {
+        return;
+    } else if (old_status == TTransactionStatus::PREPARED && new_status == TTransactionStatus::PREPARE) {
+        return;
+    } else if (old_status == TTransactionStatus::COMMITTED &&
+               (new_status != TTransactionStatus::VISIBLE && new_status != TTransactionStatus::UNKNOWN)) {
+        return;
+    }
+    _txn_state.txn_status = new_status;
+    _txn_state.reason = reason;
+}
+
+bool TxnStateHandler::_is_finished_txn_state() {
+    // The load can be successful or failed. When successful, the transaction status is VISIBLE.
+    // When failed, the transaction status can be COMMITTED, ABORTED, or UNKNOWN. COMMITTED is a
+    // special status
+    switch (_txn_state.txn_status) {
+    case TTransactionStatus::VISIBLE:
+    case TTransactionStatus::ABORTED:
+    case TTransactionStatus::UNKNOWN:
+        return true;
+    case TTransactionStatus::COMMITTED:
+        return _committed_status_from_fe;
+    default:
+        return false;
+    }
+}
+
+StatusOr<TxnState> TxnStateSubscriber::wait_finished_state(int64_t timeout_us) {
+    return _entry->value().wait_finished_state(_name, timeout_us);
+}
+
+TxnState TxnStateSubscriber::current_state() {
+    return _entry->value().txn_state();
+}
+
+inline int64_t get_current_ms() {
+    int64_t current_ts = MonotonicMillis();
+    TEST_SYNC_POINT_CALLBACK("TxnStatePoller::get_current_ms", &current_ts);
+    return current_ts;
+}
+
+Status TxnStatePoller::init() {
```

Comment:
```suggestion
void TxnStatePoller::init() {
```

Reply:
although it's always return ok currently, but I want to keep it as a style. ^-^